KOGITO-764: Create a Guided Tour for first time users on DMN editor

karreiro opened the pull request:
See https://issues.redhat.com/browse/KOGITO-764

![2020-06-17 10 08 27](https://user-images.githubusercontent.com/1079279/84905077-8dd6eb00-b086-11ea-9bad-2da7de7405bd.gif)

<img width="2011" alt="Screen Shot 2020-06-17 at 10 09 17" src="https://user-images.githubusercontent.com/1079279/84905103-962f2600-b086-11ea-8930-3265372c824f.png">

---

Part of an ensemble:
- https://github.com/kiegroup/appformer/pull/988
- https://github.com/kiegroup/kie-wb-common/pull/3336
- https://github.com/kiegroup/kogito-tooling/pull/182

---

Assumptions:
- This PR does not include i18n for any tour. It will be separately implemented by https://issues.redhat.com/browse/KOGITO-2471;
- This PR does not include the final version of the tour, and the steps are a mix of the preliminary input provided by Liz and the Learn DMN in 15 minutes (as accordingly with @ederign).

yesamer commented:
@karreiro Impressive! Congratulations for your excellent job!

jomarko commented:
### 01
When full screen is active, the guided tour is not so close tho given element
![Screenshot from 2020-06-18 19-19-31](https://user-images.githubusercontent.com/8044780/85052163-c2d36280-b198-11ea-93f5-fbb55c6f39c0.png)

jomarko commented:
### 03
When I was asked to name node 'Can drive?' was redirected to screen where help panel was shown unexpectedly. I guess because shortcut for its displaying contains '?'

I am not sure we have issue when that shortcut is active or we could just guide user to name node without '?' to be sure help panel will not unexpectedly appear. WDYT?

jomarko commented:
### 04
We should update the text of this step to be 'safe' if eager user as me already expanded properties panel somehow.
![Screenshot from 2020-06-18 19-32-19](https://user-images.githubusercontent.com/8044780/85053314-94568700-b19a-11ea-9c3c-cf058d8940ba.png)

jomarko commented:
### 05
Could we please clarify which pencil icon, I was clicking the one representing properties panel as that I interacted with in previous step
![Screenshot from 2020-06-18 19-36-58](https://user-images.githubusercontent.com/8044780/85053677-2d859d80-b19b-11ea-890c-2e0a396e334a.png)

jomarko commented:
### 06
We guide user to click on 'Select expression'. Should we mention also pressing enter as an alternative for clicking? WDYT?

jomarko commented:
### 07
I have selected an expression type but the guide seems not detected it.

jomarko commented:
### 08
We should update the text of 'Back to' step, as it is dependent on model name, not file name

![Screenshot from 2020-06-18 19-41-56](https://user-images.githubusercontent.com/8044780/85054147-e9df6380-b19b-11ea-837a-b4490d69263e.png)

jomarko commented:
### 09
Links on congratulations page for learn and doc websites should open new tabs/windows.

jomarko commented:
### 10
We have [different behavior](https://issues.redhat.com/secure/attachment/12475628/move-active.webm) chrome (first in video) and firefox (second in video)

jomarko commented:
### 11
When user doesn't satisfy next step, in this case I renamed to `Ag` instead of `Age`. He is asked if want to dismiss or continue. I decided to continue however the tour content is not centered and close buttons is missing.

![Screenshot from 2020-06-18 20-02-37](https://user-images.githubusercontent.com/8044780/85056042-d681c780-b19e-11ea-92e9-bae40f839343.png)

jomarko commented:
### 12
When user is asked to open expression editor of 'Can drive?' node, it is assumed as done if user open an expression editor of any decision. Is it expected?

tiagobento commented on packages/online-editor/src/tour/GuidedTourInitializer.tsx:
```diff
@@ -0,0 +1,102 @@
+  useEffect(() => {
+    if (isEditorReady && file.editorType === EditorType.DMN) {
+      KogitoGuidedTour.registerTutorial(getOnlineEditorTutorial());
+      KogitoGuidedTour.start("DMN Online Editor Tutorial");
```
Should this string be the same as the tutorial label? I feel you could do `tutorial.label` here 😛

karreiro replied:
Well spotted 😜

karreiro commented on packages/guided-tour/.prettierrc:
```diff
@@ -0,0 +1,3 @@
+{
+  "printWidth": 120
+}
```
Removed! :-)

karreiro commented on packages/guided-tour/package.json:
```diff
@@ -0,0 +1,40 @@
+    "prettier": "1.19.1",
+    "react": "16.12.0",
+    "react-dom": "16.12.0"
+  },
```
Fixed! Thanks!

karreiro commented on packages/guided-tour/src/components/HighlightLayer.tsx:
```diff
@@ -0,0 +1,69 @@
+      <path d={internalRectPath()} style={{ fill: "rgba(0, 0, 0, .5)" }} />
+    </svg>
+  );
+};
```
Thanks for the comment. I agree with the idea of using `useMemo` to avoid unnecessary calculations, but the `currentRefElementPosition` will almost every time hold a different value, so the amount of memoized values will be quit large. Thus, I believe we should keep `HighlightLayer` as is, but please let me know wdyt.

tiagobento replied:
When a new value comes to play, the old value is discarded and garbage-collected, so no need to worry about memory consumption.

karreiro replied:
Thanks for the explanation :-) Fixed!

karreiro commented on packages/guided-tour/src/components/NavigationControls.tsx:
```diff
@@ -0,0 +1,81 @@
+
+  function next() {
+    setCurrentStep(currentStep + 1);
+  }
```
Fixed! :-)

karreiro commented on packages/guided-tour/src/components/NavigationControls.tsx:
```diff
@@ -0,0 +1,81 @@
+      </Button>
+    </div>
+  );
+};
```
Makes sense, thanks for the explanation, @tiagobento!

karreiro commented on packages/guided-tour/src/components/utils/GuidedTourBusListeners.ts:
```diff
@@ -0,0 +1,52 @@
+function addListener(eventLabel: EventLabel, consumer: (customEvent: CustomEvent) => void) {
+  const effect = createEffect(eventLabel, consumer);
+  useLayoutEffect(effect, []);
+}
```
Renamed! Thanks!

karreiro commented on packages/guided-tour/src/components/utils/GuidedTourBusListeners.ts:
```diff
@@ -0,0 +1,52 @@
+function createEffect(eventLabel: EventLabel, consumer: (customEvent: CustomEvent) => void) {
+  function listener(e: any) {
+    consumer(e as CustomEvent);
+  }
```
Fixed!

karreiro commented on packages/guided-tour/src/components/utils/GuidedTourBusListeners.ts:
```diff
@@ -0,0 +1,52 @@
+
+function addListener(eventLabel: EventLabel, consumer: (customEvent: CustomEvent) => void) {
+  const effect = createEffect(eventLabel, consumer);
+  useLayoutEffect(effect, []);
```
Yeah, I agree, but this kind of listener must have its setup before the browser paints the change, see:

<img width="907" alt="img1" src="https://user-images.githubusercontent.com/1079279/85337748-67b2b000-b4b7-11ea-8a94-4f0005e371ca.png">
_Source: Learning React, 2nd Edition (by Alex Banks, Eve Porcello, release Date: June 2020)_

karreiro commented on packages/guided-tour/src/components/utils/GuidedTourContent.tsx:
```diff
@@ -0,0 +1,145 @@
+        </>
+      );
+    }
+  }
```
Thanks! I applied the fix on `Dialog.tsx`.

karreiro commented on packages/guided-tour/src/components/utils/GuidedTourContent.tsx:
```diff
@@ -0,0 +1,145 @@
+  );
+};
+
+export const renderNegativeReinforcementDialog = (suggestion: string, onCloseAction: () => void) => {
```
+1 for that! Thanks! :-)

karreiro commented on packages/guided-tour/src/components/utils/GuidedTourContent.tsx:
```diff
@@ -0,0 +1,145 @@
+) => {
+  const { currentStep, setCurrentStep } = useContext(CurrentTutorialContext);
+  const nextStep = () => setCurrentStep(currentStep + 1);
+  const prevStep = () => setCurrentStep(currentStep - 1);
```
Fixed!

karreiro commented on packages/guided-tour/src/components/utils/GuidedTourSelectorHandler.ts:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    setCurrentRefElementPosition(DEFAULT_RECT);
+  }
```
Fixed!

karreiro commented on packages/guided-tour/src/components/utils/GuidedTourSelectorHandler.ts:
```diff
@@ -0,0 +1,63 @@
+    CurrentTutorialContext
+  );
+  const dialogStep = getCurrentStep(currentStep, currentTutorial);
+  const selector = dialogStep?.selector || "";
```
Fixed!

karreiro commented on packages/guided-tour/src/components/utils/GuidedTourStepsHelper.ts:
```diff
@@ -0,0 +1,44 @@
+  return getSteps(currentTutorial)[currentStep];
+};
+
+export const getSteps = (currentTutorial: Tutorial | undefined) => {
```
Fixed! Thanks!

karreiro commented on packages/guided-tour/src/components/utils/GuidedTourStepsHelper.ts:
```diff
@@ -0,0 +1,44 @@
+      return tutorial.label === stepTutorialLabel;
+    });
+
+    return stepTutorial?.steps || [];
```
Fixed! (I'm keeping it only on `HighlightLayer`, because there it has a logical operation (since I want to return the right-side when the left-side is `false`).

karreiro commented on packages/guided-tour/src/index.tsx:
```diff
@@ -0,0 +1,154 @@
+import { GuidedTour } from "./components";
+import { GuidedTourDomUtils, GuidedTourEventBus, GuidedTourCookie } from "./core";
+
+class KogitoGuidedTour {
```
Yeah! It's a singleton now!

karreiro commented on packages/microeditor-envelope-protocol/src/EnvelopeBusOuterMessageHandler.ts:
```diff
@@ -41,6 +42,9 @@ export interface EnvelopeBusOuterMessageHandlerImpl {
   receive_newEdit(edit: KogitoEdit): void;
+  receive_guidedTourUserInteraction(userInteraction: UserInteraction): void;
+  receive_guidedTourRegisterTutorial(tutorial: Tutorial): void;
+  receive_guidedTourElementPositionRequest(rect: Rect): void;
```
Renamed!

karreiro commented on packages/microeditor-envelope/src/api/tour/GuidedTourService.ts:
```diff
@@ -0,0 +1,25 @@
+/**
+ * PUBLIC ENVELOPE API
+ */
+export interface GuidedTourService {
```
Renamed!

karreiro commented on packages/online-editor/src/App.tsx:
```diff
@@ -72,12 +74,16 @@ export function App(props: Props) {
     [file]
   );
 
+  useDMNTour(isEditorReady, file);
```
Renamed!

karreiro commented on packages/online-editor/src/common/GlobalContext.ts:
```diff
@@ -25,6 +25,8 @@ export interface GlobalContextType {
   readonly: boolean;
   external: boolean;
+  isEditorReady: boolean;
+  setIsEditorReady: (isEditorReady: boolean) => void;
```
Fixed! Good catch, thanks! :-)

karreiro commented:
@tiagobento I've answered your comments inline, thank you for the valuable tips.

----

@jomarko Thank you for your review and the reported issues. Regarding each item you've raised:

- 1 - I believe we can handle this fullscreen bug separately here: https://issues.redhat.com/browse/KOGITO-2513

- 2 - So... it will be handled by [KOGITO-2513](https://issues.redhat.com/browse/KOGITO-2513)

- 3 - Yeah, this is an issue, but it's being fixed by Tiago here: https://github.com/kiegroup/kogito-tooling/pull/185

- 4 - Makes sense... Fixed!

- 5 - Well spotted :-) Fixed!

- 6 - That's a good idea. Fixed!

- 7 - I've added some enhancements on the grid observer, and I couldn't reproduce this issue after that. Before, these changes, I could intermittently reproduce it. So, I believe it's fixed! :-)

- 8 - We would need extra infrastructure code to get this kind of information. Maybe we can have that kind of infra in the future, but for now I believe we can just point the link by using the dialog.

- 9 - Fixed!

- 10 - Well spotted. Fixed!

- 11 - Makes sense :-) Fixed!

- 12 - Well spotted, thanks. They are fixed now.

karreiro commented:
(conflicts fixed)

karreiro commented:
(conflicts fixed) 👀

karreiro commented:
Hi @jomarko, I'm not sure if you could solve the problems to build this ensemble locally, but you can test this PR by following these steps:
- Open `packages/kie-bc-editors-unpacked/dmn/`
- Unzip [this zip file](https://www.dropbox.com/s/b5c4d4osyhl45yv/dmn.zip?dl=0)
- Now, on `packages/online-editor` you can run `yarn run start` and it will work
I hope this is helpful :-)

jomarko commented:
### Re-review 01
New issue reported https://issues.redhat.com/browse/KOGITO-2602

jomarko commented:
### Re-review 02
After user select some expression type the tour step provides link for 'learn-in-15-minutes'. This should be also opened in new tab. see [video](https://issues.redhat.com/secure/attachment/12476527/learn-in-15-minutes.webm).

jomarko commented:
### Re-review 03
New issue reported https://issues.redhat.com/browse/KOGITO-2603

jomarko commented:
### Re-review 04
New issue reported https://issues.redhat.com/browse/KOGITO-2605

jomarko commented on packages/guided-tour/src/__tests__/components/NavigationControls.test.tsx:
```diff
@@ -0,0 +1,122 @@
+      currentTutorial: { steps: [{}, {}, {}, {}] }
+    };
+
+    act(() => {
+      useContextMock(ctx);
+      render(<NavigationControls />);
+
+      triggerClick("[data-kgt-next]");
+    });
```
I am not able to manually switch between steps. Is this  test for tour that will be launched in other mode?

karreiro replied:
Yeah, currently none of the current steps is enabling button-based navigation. This is something we already support, but this feature is not being used on the current steps.

karreiro commented:
Thank you for the comments, @jomarko. May I handle the “Re-review 2” in a separate JIRA as well? I can open another PR by the EOD. But I would like to get this ensemble in :-)

Thanks again for the comments. Well spotted issues.
